"Enhancement: forward refs" actually broke forward refs

After #69 was merged, forward refs are not parsed correctly anymore.  
It complains that the type, which is forward referred to, is not defined.

Code:
```
@frozen_dataclass
class Comment:
    ...
    replies: List['Comment']
```

Error: 
`
pedantic.exceptions.PedanticTypeCheckException: In dataclass "Comment" in field "replies": An error occurred during type hint checking. Value: [Comment(...)] Annotation: typing.List[ForwardRef('Comment')] Mostly this is caused by an incorrect type annotation. Details: name 'Comment' is not defined
`

Ah, I missed a crucial part of the stacktrace (which is quite long and confusing).  
The problem occured in a `copy_with` statement, I think.  
Could it be the copy_with implementation does not correctly handle ForwardRefs?

```
type_ = 'Comment', globals_ = None
context = {'Any': typing.Any, 'Callable': typing.Callable, 'Dict': typing.Dict, 'T': ~T, ...}

    def resolve_forward_ref(type_: str, globals_: Dict[str, Any] = None, context: Dict = None) -> Type:
        """
            Resolve a type annotation that is a string.
    
            Raises:
                NameError: in case of [type_] cannot be resolved.
        """
    
>       return eval(str(type_), globals_ or globals(), context or {})

.../lib/python3.10/site-packages/pedantic/type_checking_logic/resolve_forward_ref.py:12: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   NameError: name 'Comment' is not defined

<string>:1: NameError

During handling of the above exception, another exception occurred:
   
>       comment_1 = comment_1.copy_with(replies=[reply_1])

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.../lib/python3.10/site-packages/pedantic/decorators/cls_deco_frozen_dataclass.py:81: in copy_with
    return cls_(**{**current_values, **kwargs})
<string>:14: in __init__
    ???
.../lib/python3.10/site-packages/pedantic/decorators/cls_deco_frozen_dataclass.py:124: in new_post_init
    self.validate_types(_context=context)
.../lib/python3.10/site-packages/pedantic/decorators/cls_deco_frozen_dataclass.py:105: in validate_types
    assert_value_matches_type(
.../lib/python3.10/site-packages/pedantic/type_checking_logic/check_types.py:24: in assert_value_matches_type
    if not _check_type(value=value, type_=type_, err=err, type_vars=type_vars, context=context):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

value = [Comment(...)]
type_ = typing.List[ForwardRef('Comment')]
err = 'In dataclass "Comment" in field "replies": ', type_vars = {}
context = {'Any': typing.Any, 'Callable': typing.Callable, 'Dict': typing.Dict, 'T': ~T, ...}

    def _check_type(value: Any, type_: Any, err: str, type_vars: Dict[TypeVar_, Any], context: Dict[str, Any] = None) -> bool:    
        if type_ is None:
            return value == type_
        elif isinstance(type_, str):
            class_name = value.__class__.__name__
            base_class_name = value.__class__.__base__.__name__
            return class_name == type_ or base_class_name == type_
    
        if isinstance(type_, tuple):
            raise PedanticTypeCheckException(f'{err}Use "Tuple[]" instead of "{type_}" as type hint.')
        if isinstance(type_, list):
            raise PedanticTypeCheckException(f'{err}Use "List[]" instead of "{type_}" as type hint.')
        if type_ is tuple:
            raise PedanticTypeCheckException(f'{err}Use "Tuple[]" instead of "tuple" as type hint.')
        if type_ is list:
            raise PedanticTypeCheckException(f'{err}Use "List[]" instead of "list" as type hint.')
        if type_ is dict:
            raise PedanticTypeCheckException(f'{err}Use "Dict[]" instead of "dict" as type hint.')
        if type_ is set:
            raise PedanticTypeCheckException(f'{err}Use "Set[]" instead of "set" as type hint.')
        if type_ is frozenset:
            raise PedanticTypeCheckException(f'{err}Use "FrozenSet[]" instead of "frozenset" as type hint.')
        if type_ is type:
            raise PedanticTypeCheckException(f'{err}Use "Type[]" instead of "type" as type hint.')
    
        try:
            return _is_instance(obj=value, type_=type_, type_vars=type_vars, context=context)
        except PedanticTypeCheckException as ex:
            raise PedanticTypeCheckException(f'{err} {ex}')
        except PedanticTypeVarMismatchException as ex:
            raise PedanticTypeVarMismatchException(f'{err} {ex}')
        except (AttributeError, Exception) as ex:
>           raise PedanticTypeCheckException(
                f'{err}An error occurred during type hint checking. Value: {value} Annotation: '
                f'{type_} Mostly this is caused by an incorrect type annotation. Details: {ex} ')
E           pedantic.exceptions.PedanticTypeCheckException: In dataclass "Comment" in field "replies": An error occurred during type hint checking. Value: [Comment(...)] Annotation: typing.List[ForwardRef('Comment')] Mostly this is caused by an incorrect type annotation. Details: name 'Comment' is not defined

.../lib/python3.10/site-packages/pedantic/type_checking_logic/check_types.py:136: PedanticTypeCheckException
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,4 +1,7 @@
 # Changelog
+## Pedantic 1.13.2
+- [fixes a bug in the type-checking of `frozen_dataclass` and `frozen_type_safe_dataclass` in context of `ForwardRef`s](https://github.com/LostInDarkMath/pedantic-python-decorators/issues/72)
+
 ## Pedantic 1.13.1
 - fix `ImportError` of optional `multiprocess` package
 - fix typo in `README.md`

--- a/docs/pedantic/decorators/cls_deco_frozen_dataclass.html
+++ b/docs/pedantic/decorators/cls_deco_frozen_dataclass.html
@@ -148,7 +148,10 @@ def frozen_dataclass(
 
             def new_post_init(self) -&gt; None:
                 old_post_init(self)
-                context = get_context(3)
+                context = get_context(depth=3, increase_depth_if_name_matches=[
+                    copy_with.__name__,
+                    deep_copy_with.__name__,
+                ])
                 self.validate_types(_context=context)
 
             setattr(cls_, &#39;__post_init__&#39;, new_post_init)
@@ -333,7 +336,10 @@ Foo(a=676676, b='new', c=False)
 
             def new_post_init(self) -&gt; None:
                 old_post_init(self)
-                context = get_context(3)
+                context = get_context(depth=3, increase_depth_if_name_matches=[
+                    copy_with.__name__,
+                    deep_copy_with.__name__,
+                ])
                 self.validate_types(_context=context)
 
             setattr(cls_, &#39;__post_init__&#39;, new_post_init)

--- a/docs/pedantic/decorators/fn_deco_pedantic.html
+++ b/docs/pedantic/decorators/fn_deco_pedantic.html
@@ -29,6 +29,7 @@
 <pre><code class="python">from functools import wraps
 from typing import Any, Optional
 
+from pedantic.get_context import get_context
 from pedantic.type_checking_logic.check_docstring import _check_docstring
 from pedantic.constants import ReturnType, F
 from pedantic.models.decorated_function import DecoratedFunction
@@ -73,28 +74,23 @@ def pedantic(func: Optional[F] = None, require_docstring: bool = False) -&gt; F:
        Use kwargs when you call function my_function. Args: (5, 4.0, &#39;hi&#39;)
    &#34;&#34;&#34;
 
-    import sys
-    frame = sys._getframe(1)  # get parent frame
-
     def decorator(f: F) -&gt; F:
-        nonlocal frame
-
         if not is_enabled():
             return f
 
-        decorated_func = DecoratedFunction(func=f, context={**frame.f_globals, **frame.f_locals})
+        decorated_func = DecoratedFunction(func=f)
 
         if decorated_func.docstring is not None and (require_docstring or len(decorated_func.docstring.params)) &gt; 0:
             _check_docstring(decorated_func=decorated_func)
 
         @wraps(f)
         def wrapper(*args: Any, **kwargs: Any) -&gt; ReturnType:
-            call = FunctionCall(func=decorated_func, args=args, kwargs=kwargs)
+            call = FunctionCall(func=decorated_func, args=args, kwargs=kwargs, context=get_context(2))
             call.assert_uses_kwargs()
             return call.check_types()
 
         async def async_wrapper(*args: Any, **kwargs: Any) -&gt; ReturnType:
-            call = FunctionCall(func=decorated_func, args=args, kwargs=kwargs)
+            call = FunctionCall(func=decorated_func, args=args, kwargs=kwargs, context=get_context(2))
             call.assert_uses_kwargs()
             return await call.async_check_types()
 
@@ -200,28 +196,23 @@ Use kwargs when you call function my_function. Args: (5, 4.0, 'hi')
        Use kwargs when you call function my_function. Args: (5, 4.0, &#39;hi&#39;)
    &#34;&#34;&#34;
 
-    import sys
-    frame = sys._getframe(1)  # get parent frame
-
     def decorator(f: F) -&gt; F:
-        nonlocal frame
-
         if not is_enabled():
             return f
 
-        decorated_func = DecoratedFunction(func=f, context={**frame.f_globals, **frame.f_locals})
+        decorated_func = DecoratedFunction(func=f)
 
         if decorated_func.docstring is not None and (require_docstring or len(decorated_func.docstring.params)) &gt; 0:
             _check_docstring(decorated_func=decorated_func)
 
         @wraps(f)
         def wrapper(*args: Any, **kwargs: Any) -&gt; ReturnType:
-            call = FunctionCall(func=decorated_func, args=args, kwargs=kwargs)
+            call = FunctionCall(func=decorated_func, args=args, kwargs=kwargs, context=get_context(2))
             call.assert_uses_kwargs()
             return call.check_types()
 
         async def async_wrapper(*args: Any, **kwargs: Any) -&gt; ReturnType:
-            call = FunctionCall(func=decorated_func, args=args, kwargs=kwargs)
+            call = FunctionCall(func=decorated_func, args=args, kwargs=kwargs, context=get_context(2))
             call.assert_uses_kwargs()
             return await call.async_check_types()
 

--- a/docs/pedantic/decorators/fn_deco_require_kwargs.html
+++ b/docs/pedantic/decorators/fn_deco_require_kwargs.html
@@ -55,7 +55,7 @@ def require_kwargs(func: F) -&gt; F:
     @wraps(func)
     def wrapper(*args: Any, **kwargs: Any) -&gt; ReturnType:
         decorated_func = DecoratedFunction(func=func)
-        call = FunctionCall(func=decorated_func, args=args, kwargs=kwargs)
+        call = FunctionCall(func=decorated_func, args=args, kwargs=kwargs, context={})
         call.assert_uses_kwargs()
         return func(*args, **kwargs)
     return wrapper
@@ -115,7 +115,7 @@ Use kwargs when you call function my_function. Args: (5, 4, 3)
     @wraps(func)
     def wrapper(*args: Any, **kwargs: Any) -&gt; ReturnType:
         decorated_func = DecoratedFunction(func=func)
-        call = FunctionCall(func=decorated_func, args=args, kwargs=kwargs)
+        call = FunctionCall(func=decorated_func, args=args, kwargs=kwargs, context={})
         call.assert_uses_kwargs()
         return func(*args, **kwargs)
     return wrapper</code></pre>

--- a/docs/pedantic/get_context.html
+++ b/docs/pedantic/get_context.html
@@ -27,16 +27,21 @@
 <span>Expand source code</span>
 </summary>
 <pre><code class="python">import sys
-from typing import Type, Dict
+from typing import Type, Dict, List
 
 
-def get_context(depth: int = 1) -&gt; Dict[str, Type]:
+def get_context(depth: int = 1, increase_depth_if_name_matches: List[str] = None) -&gt; Dict[str, Type]:
     &#34;&#34;&#34;
         Get the context of a frame at the given depth of the current call stack.
         See also: https://docs.python.org/3/library/sys.html#sys._getframe
     &#34;&#34;&#34;
 
     frame = sys._getframe(depth)
+    name = frame.f_code.co_name
+
+    if name in (increase_depth_if_name_matches or []):
+        frame = sys._getframe(depth + 1)
+
     return {**frame.f_globals, **frame.f_locals}</code></pre>
 </details>
 </section>
@@ -48,7 +53,7 @@ def get_context(depth: int = 1) -&gt; Dict[str, Type]:
 <h2 class="section-title" id="header-functions">Functions</h2>
 <dl>
 <dt id="pedantic.get_context.get_context"><code class="name flex">
-<span>def <span class="ident">get_context</span></span>(<span>depth: int = 1) ‑> Dict[str, Type[+CT_co]]</span>
+<span>def <span class="ident">get_context</span></span>(<span>depth: int = 1, increase_depth_if_name_matches: List[str] = None) ‑> Dict[str, Type[+CT_co]]</span>
 </code></dt>
 <dd>
 <div class="desc"><p>Get the context of a frame at the given depth of the current call stack.
@@ -57,13 +62,18 @@ See also: <a href="https://docs.python.org/3/library/sys.html#sys._getframe">htt
 <summary>
 <span>Expand source code</span>
 </summary>
-<pre><code class="python">def get_context(depth: int = 1) -&gt; Dict[str, Type]:
+<pre><code class="python">def get_context(depth: int = 1, increase_depth_if_name_matches: List[str] = None) -&gt; Dict[str, Type]:
     &#34;&#34;&#34;
         Get the context of a frame at the given depth of the current call stack.
         See also: https://docs.python.org/3/library/sys.html#sys._getframe
     &#34;&#34;&#34;
 
     frame = sys._getframe(depth)
+    name = frame.f_code.co_name
+
+    if name in (increase_depth_if_name_matches or []):
+        frame = sys._getframe(depth + 1)
+
     return {**frame.f_globals, **frame.f_locals}</code></pre>
 </details>
 </dd>

--- a/docs/pedantic/models/decorated_function.html
+++ b/docs/pedantic/models/decorated_function.html
@@ -48,9 +48,8 @@ FUNCTIONS_THAT_REQUIRE_KWARGS = [
 
 
 class DecoratedFunction:
-    def __init__(self, func: Callable[..., Any], context: Dict[str, Any] = None) -&gt; None:
+    def __init__(self, func: Callable[..., Any]) -&gt; None:
         self._func = func
-        self._context = context
 
         if not isinstance(func, (types.FunctionType, types.MethodType)):
             raise PedanticTypeCheckException(f&#39;{self.full_name} should be a method or function.&#39;)
@@ -68,10 +67,6 @@ class DecoratedFunction:
     @property
     def func(self) -&gt; Callable[..., Any]:
         return self._func
-
-    @property
-    def context(self) -&gt; Dict[str, Any]:
-        return self._context
 
     @property
     def annotations(self) -&gt; Dict[str, Any]:
@@ -162,7 +157,7 @@ class DecoratedFunction:
 <dl>
 <dt id="pedantic.models.decorated_function.DecoratedFunction"><code class="flex name class">
 <span>class <span class="ident">DecoratedFunction</span></span>
-<span>(</span><span>func: Callable[..., Any], context: Dict[str, Any] = None)</span>
+<span>(</span><span>func: Callable[..., Any])</span>
 </code></dt>
 <dd>
 <div class="desc"></div>
@@ -171,9 +166,8 @@ class DecoratedFunction:
 <span>Expand source code</span>
 </summary>
 <pre><code class="python">class DecoratedFunction:
-    def __init__(self, func: Callable[..., Any], context: Dict[str, Any] = None) -&gt; None:
+    def __init__(self, func: Callable[..., Any]) -&gt; None:
         self._func = func
-        self._context = context
 
         if not isinstance(func, (types.FunctionType, types.MethodType)):
             raise PedanticTypeCheckException(f&#39;{self.full_name} should be a method or function.&#39;)
@@ -191,10 +185,6 @@ class DecoratedFunction:
     @property
     def func(self) -&gt; Callable[..., Any]:
         return self._func
-
-    @property
-    def context(self) -&gt; Dict[str, Any]:
-        return self._context
 
     @property
     def annotations(self) -&gt; Dict[str, Any]:
@@ -285,18 +275,6 @@ class DecoratedFunction:
 <pre><code class="python">@property
 def annotations(self) -&gt; Dict[str, Any]:
     return self._full_arg_spec.annotations</code></pre>
-</details>
-</dd>
-<dt id="pedantic.models.decorated_function.DecoratedFunction.context"><code class="name">var <span class="ident">context</span> : Dict[str, Any]</code></dt>
-<dd>
-<div class="desc"></div>
-<details class="source">
-<summary>
-<span>Expand source code</span>
-</summary>
-<pre><code class="python">@property
-def context(self) -&gt; Dict[str, Any]:
-    return self._context</code></pre>
 </details>
 </dd>
 <dt id="pedantic.models.decorated_function.DecoratedFunction.docstring"><code class="name">var <span class="ident">docstring</span> : Optional[docstring_parser.common.Docstring]</code></dt>
@@ -535,7 +513,6 @@ def wants_args(self) -&gt; bool:
 <h4><code><a title="pedantic.models.decorated_function.DecoratedFunction" href="#pedantic.models.decorated_function.DecoratedFunction">DecoratedFunction</a></code></h4>
 <ul class="two-column">
 <li><code><a title="pedantic.models.decorated_function.DecoratedFunction.annotations" href="#pedantic.models.decorated_function.DecoratedFunction.annotations">annotations</a></code></li>
-<li><code><a title="pedantic.models.decorated_function.DecoratedFunction.context" href="#pedantic.models.decorated_function.DecoratedFunction.context">context</a></code></li>
 <li><code><a title="pedantic.models.decorated_function.DecoratedFunction.docstring" href="#pedantic.models.decorated_function.DecoratedFunction.docstring">docstring</a></code></li>
 <li><code><a title="pedantic.models.decorated_function.DecoratedFunction.err" href="#pedantic.models.decorated_function.DecoratedFunction.err">err</a></code></li>
 <li><code><a title="pedantic.models.decorated_function.DecoratedFunction.full_name" href="#pedantic.models.decorated_function.DecoratedFunction.full_name">full_name</a></code></li>

--- a/docs/pedantic/models/function_call.html
+++ b/docs/pedantic/models/function_call.html
@@ -27,7 +27,7 @@
 <span>Expand source code</span>
 </summary>
 <pre><code class="python">import inspect
-from typing import Dict, Tuple, Any, Union
+from typing import Dict, Tuple, Any, Union, Type
 
 from pedantic.constants import TypeVar, TYPE_VAR_METHOD_NAME, ReturnType
 from pedantic.exceptions import PedanticCallWithArgsException, PedanticTypeCheckException
@@ -37,10 +37,11 @@ from pedantic.models.generator_wrapper import GeneratorWrapper
 
 
 class FunctionCall:
-    def __init__(self, func: DecoratedFunction, args: Tuple[Any, ...], kwargs: Dict[str, Any]) -&gt; None:
+    def __init__(self, func: DecoratedFunction, args: Tuple[Any, ...], kwargs: Dict[str, Any], context: Dict[str, Type]) -&gt; None:
         self._func = func
         self._args = args
         self._kwargs = kwargs
+        self._context = context
         self._instance = self.args[0] if self.func.is_instance_method else None
         self._type_vars = dict()
         self._params_without_self = {k: v for k, v in self.func.signature.parameters.items() if v.name != &#39;self&#39;}
@@ -129,7 +130,7 @@ class FunctionCall:
                 err=self.func.err,
                 type_vars=self.type_vars,
                 key=key,
-                context=self.func.context,
+                context=self._context,
             )
 
     def _check_types_args(self, params: Dict[str, inspect.Parameter]) -&gt; None:
@@ -144,7 +145,7 @@ class FunctionCall:
                 type_=expected,
                 err=self.func.err,
                 type_vars=self.type_vars,
-                context=self.func.context,
+                context=self._context,
             )
 
     def _check_types_kwargs(self, params: Dict[str, inspect.Parameter]) -&gt; None:
@@ -162,7 +163,7 @@ class FunctionCall:
                 err=self.func.err,
                 type_vars=self.type_vars,
                 key=kwarg,
-                context=self.func.context,
+                context=self._context,
             )
 
     def _check_types_return(self, result: Any) -&gt; Union[None, GeneratorWrapper]:
@@ -184,7 +185,7 @@ class FunctionCall:
             err=self.func.err,
             type_vars=self.type_vars,
             msg=msg,
-            context=self.func.context,
+            context=self._context,
         )
         return result
 
@@ -216,7 +217,7 @@ class FunctionCall:
 <dl>
 <dt id="pedantic.models.function_call.FunctionCall"><code class="flex name class">
 <span>class <span class="ident">FunctionCall</span></span>
-<span>(</span><span>func: <a title="pedantic.models.decorated_function.DecoratedFunction" href="decorated_function.html#pedantic.models.decorated_function.DecoratedFunction">DecoratedFunction</a>, args: Tuple[Any, ...], kwargs: Dict[str, Any])</span>
+<span>(</span><span>func: <a title="pedantic.models.decorated_function.DecoratedFunction" href="decorated_function.html#pedantic.models.decorated_function.DecoratedFunction">DecoratedFunction</a>, args: Tuple[Any, ...], kwargs: Dict[str, Any], context: Dict[str, Type[+CT_co]])</span>
 </code></dt>
 <dd>
 <div class="desc"></div>
@@ -225,10 +226,11 @@ class FunctionCall:
 <span>Expand source code</span>
 </summary>
 <pre><code class="python">class FunctionCall:
-    def __init__(self, func: DecoratedFunction, args: Tuple[Any, ...], kwargs: Dict[str, Any]) -&gt; None:
+    def __init__(self, func: DecoratedFunction, args: Tuple[Any, ...], kwargs: Dict[str, Any], context: Dict[str, Type]) -&gt; None:
         self._func = func
         self._args = args
         self._kwargs = kwargs
+        self._context = context
         self._instance = self.args[0] if self.func.is_instance_method else None
         self._type_vars = dict()
         self._params_without_self = {k: v for k, v in self.func.signature.parameters.items() if v.name != &#39;self&#39;}
@@ -317,7 +319,7 @@ class FunctionCall:
                 err=self.func.err,
                 type_vars=self.type_vars,
                 key=key,
-                context=self.func.context,
+                context=self._context,
             )
 
     def _check_types_args(self, params: Dict[str, inspect.Parameter]) -&gt; None:
@@ -332,7 +334,7 @@ class FunctionCall:
                 type_=expected,
                 err=self.func.err,
                 type_vars=self.type_vars,
-                context=self.func.context,
+                context=self._context,
             )
 
     def _check_types_kwargs(self, params: Dict[str, inspect.Parameter]) -&gt; None:
@@ -350,7 +352,7 @@ class FunctionCall:
                 err=self.func.err,
                 type_vars=self.type_vars,
                 key=kwarg,
-                context=self.func.context,
+                context=self._context,
             )
 
     def _check_types_return(self, result: Any) -&gt; Union[None, GeneratorWrapper]:
@@ -372,7 +374,7 @@ class FunctionCall:
             err=self.func.err,
             type_vars=self.type_vars,
             msg=msg,
-            context=self.func.context,
+            context=self._context,
         )
         return result
 

--- a/docs/pedantic/tests/test_frozen_dataclass.html
+++ b/docs/pedantic/tests/test_frozen_dataclass.html
@@ -361,7 +361,18 @@ class TestFrozenDataclass(unittest.TestCase):
         with self.assertRaises(expected_exception=PedanticTypeCheckException):
             StateChangeResult(new_state=OnlineMachineState())
 
-        s.validate_types()</code></pre>
+        s.validate_types()
+
+    def test_forward_ref_to_itself(self):
+        &#34;&#34;&#34; Regression test for https://github.com/LostInDarkMath/pedantic-python-decorators/issues/72 &#34;&#34;&#34;
+
+        @frozen_type_safe_dataclass
+        class Comment:
+            replies: List[&#39;Comment&#39;]
+
+        comment = Comment(replies=[Comment(replies=[])])
+        comment.copy_with(replies=[Comment(replies=[])])
+        comment.validate_types()</code></pre>
 </details>
 </section>
 <section>
@@ -1025,7 +1036,18 @@ not have a method with the specified name.</p></div>
         with self.assertRaises(expected_exception=PedanticTypeCheckException):
             StateChangeResult(new_state=OnlineMachineState())
 
-        s.validate_types()</code></pre>
+        s.validate_types()
+
+    def test_forward_ref_to_itself(self):
+        &#34;&#34;&#34; Regression test for https://github.com/LostInDarkMath/pedantic-python-decorators/issues/72 &#34;&#34;&#34;
+
+        @frozen_type_safe_dataclass
+        class Comment:
+            replies: List[&#39;Comment&#39;]
+
+        comment = Comment(replies=[Comment(replies=[])])
+        comment.copy_with(replies=[Comment(replies=[])])
+        comment.validate_types()</code></pre>
 </details>
 <h3>Ancestors</h3>
 <ul class="hlist">
@@ -1138,6 +1160,27 @@ not have a method with the specified name.</p></div>
 
     self.assertNotEqual(a, c)
     self.assertNotEqual(hash(a), hash(c))</code></pre>
+</details>
+</dd>
+<dt id="pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_forward_ref_to_itself"><code class="name flex">
+<span>def <span class="ident">test_forward_ref_to_itself</span></span>(<span>self)</span>
+</code></dt>
+<dd>
+<div class="desc"><p>Regression test for <a href="https://github.com/LostInDarkMath/pedantic-python-decorators/issues/72">https://github.com/LostInDarkMath/pedantic-python-decorators/issues/72</a></p></div>
+<details class="source">
+<summary>
+<span>Expand source code</span>
+</summary>
+<pre><code class="python">def test_forward_ref_to_itself(self):
+    &#34;&#34;&#34; Regression test for https://github.com/LostInDarkMath/pedantic-python-decorators/issues/72 &#34;&#34;&#34;
+
+    @frozen_type_safe_dataclass
+    class Comment:
+        replies: List[&#39;Comment&#39;]
+
+    comment = Comment(replies=[Comment(replies=[])])
+    comment.copy_with(replies=[Comment(replies=[])])
+    comment.validate_types()</code></pre>
 </details>
 </dd>
 <dt id="pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_frozen_dataclass_above_dataclass"><code class="name flex">
@@ -1566,6 +1609,7 @@ not have a method with the specified name.</p></div>
 <li><code><a title="pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_copy_with_and_deep_copy_with" href="#pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_copy_with_and_deep_copy_with">test_copy_with_and_deep_copy_with</a></code></li>
 <li><code><a title="pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_copy_with_is_shallow" href="#pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_copy_with_is_shallow">test_copy_with_is_shallow</a></code></li>
 <li><code><a title="pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_equals_and_hash" href="#pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_equals_and_hash">test_equals_and_hash</a></code></li>
+<li><code><a title="pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_forward_ref_to_itself" href="#pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_forward_ref_to_itself">test_forward_ref_to_itself</a></code></li>
 <li><code><a title="pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_frozen_dataclass_above_dataclass" href="#pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_frozen_dataclass_above_dataclass">test_frozen_dataclass_above_dataclass</a></code></li>
 <li><code><a title="pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_frozen_dataclass_below_dataclass" href="#pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_frozen_dataclass_below_dataclass">test_frozen_dataclass_below_dataclass</a></code></li>
 <li><code><a title="pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_frozen_dataclass_inheritance_not_override_post_init" href="#pedantic.tests.test_frozen_dataclass.TestFrozenDataclass.test_frozen_dataclass_inheritance_not_override_post_init">test_frozen_dataclass_inheritance_not_override_post_init</a></code></li>

--- a/pedantic/decorators/cls_deco_frozen_dataclass.py
+++ b/pedantic/decorators/cls_deco_frozen_dataclass.py
@@ -120,7 +120,10 @@ def frozen_dataclass(
 
             def new_post_init(self) -> None:
                 old_post_init(self)
-                context = get_context(3)
+                context = get_context(depth=3, increase_depth_if_name_matches=[
+                    copy_with.__name__,
+                    deep_copy_with.__name__,
+                ])
                 self.validate_types(_context=context)
 
             setattr(cls_, '__post_init__', new_post_init)

--- a/pedantic/get_context.py
+++ b/pedantic/get_context.py
@@ -1,12 +1,17 @@
 import sys
-from typing import Type, Dict
+from typing import Type, Dict, List
 
 
-def get_context(depth: int = 1) -> Dict[str, Type]:
+def get_context(depth: int = 1, increase_depth_if_name_matches: List[str] = None) -> Dict[str, Type]:
     """
         Get the context of a frame at the given depth of the current call stack.
         See also: https://docs.python.org/3/library/sys.html#sys._getframe
     """
 
     frame = sys._getframe(depth)
+    name = frame.f_code.co_name
+
+    if name in (increase_depth_if_name_matches or []):
+        frame = sys._getframe(depth + 1)
+
     return {**frame.f_globals, **frame.f_locals}

--- a/pedantic/tests/test_frozen_dataclass.py
+++ b/pedantic/tests/test_frozen_dataclass.py
@@ -334,3 +334,14 @@ class TestFrozenDataclass(unittest.TestCase):
             StateChangeResult(new_state=OnlineMachineState())
 
         s.validate_types()
+
+    def test_forward_ref_to_itself(self):
+        """ Regression test for https://github.com/LostInDarkMath/pedantic-python-decorators/issues/72 """
+
+        @frozen_type_safe_dataclass
+        class Comment:
+            replies: List['Comment']
+
+        comment = Comment(replies=[Comment(replies=[])])
+        comment.copy_with(replies=[Comment(replies=[])])
+        comment.validate_types()

--- a/setup.py
+++ b/setup.py
@@ -14,7 +14,7 @@ author = "Willi Sontopski"
 
 setup(
     name="pedantic",
-    version="1.13.1",
+    version="1.13.2",
     python_requires='>=3.7.0',
     packages=find_packages(),
     install_requires=[],

